cupy/core/_routines_statistics.pyx

```diff
@@ -310,6 +322,234 @@ cdef _nanargmax_func = create_reduction_func(
+        ptrdiff_t j;
+        for (j = k + id + z; j < m - (m - z) % 32; j += 32) {
+            if (a[j] < a[k - 1 + z]) {
+                ${dtype} tmp = a[k + 32 * x + id + z];
+                a[k + 32 * x + id + z] = a[j];
+                a[j] = tmp;
+                ++x;
+            }
+
+            // If at least one thread in the warp has found t values that
+            // can be selected, we update the first k elements.
+    #if __CUDACC_VER_MAJOR__ >= 9
+            if (__any_sync(0xffffffff, x >= t)) {
+    #else
+            if (__any(x >= t)) {
+    #endif
+                bitonic_sort<${dtype}>(a, k + z, 32 * t + k + z, id);
+                merge<${dtype}>(a, k, id, z, k + z, min(k, 32 * t));
+                x = 0;
+            }
+        }
+        if (j < m && a[j] < a[k - 1 + z]) {
+            ${dtype} tmp = a[k + 32 * x + id + z];
+            a[k + 32 * x + id + z] = a[j];
+            a[j] = tmp;
+        }
+
+        // Finally, we merge the first k elements and the remainders to be
+        // stored.
+        bitonic_sort<${dtype}>(a, k + z, 32 * t + k + z, id);
+        merge<${dtype}>(a, k, id, z, k + z, min(k, 32 * t));
+
+        int kth = len / 2;
+        if (len % 2)
+            out[i / 32] = a[z + kth];
+        else
+            out[i / 32] = (a[z + kth - 1] + a[z + kth]) / 2;
+    }
+    }
+    ''').substitute(name=name, dtype=dtype)
+    module = compile_with_cache(source)
+    return module.get_function(name)
+
+
+cdef _median_partition_core(ndarray data, max_k, ndarray out, int axis):
+
+    if data.dtype.kind == 'c':
+        raise NotImplementedError('Sorting arrays with dtype \'{}\' is '
+                                  'not supported'.format(data.dtype))
+
+    cdef int ndim = data._shape.size()
+    cdef Py_ssize_t length, s, sz, t
+
+    if not data._c_contiguous:
+        raise NotImplementedError('Sorting non-contiguous array is not '
+                                  'supported.')
+
+    if axis < 0:
+        axis += ndim
+    if not (0 <= axis < ndim):
+        raise numpy.AxisError('Axis out of range')
+
+    if axis != ndim - 1:
+        data = _manipulation.rollaxis(data, axis, ndim).copy()
+
+    length = data._shape[ndim-1]
+
+    # For simplicity, max_k is round up to the power of 2. If max_k is
+    # already the power of 2, it is round up to the next power of 2 because
+    # we need to collect the first max(kth)+1 elements.
+    max_k = max(32, 1 << max_k.bit_length())
+
+    # The parameter t is the length of the list that stores elements to be
+    # selected for each thread. We divide the array into sz subarrays.
+    # These parameters are determined from the measurement on TITAN X.
+    t = 4
+    sz = 1
+    while sz > 0 and length // sz < max_k + 32 * t:
+        sz //= 2
+    sz *= data.size // length
+
+    # If the array size is small or k is large, we simply sort the array.
+    if length < 32 or sz < 1 or max_k >= 1024:
+        # kth is ignored.
```

Comment:
Have you tried to always launch your kernel for smaller array sizes?

Reply:
Do you mean to say to launch the kernel for small sizes and then merge them together to fully sort the array?

Reply:
well, I guess its ok.

Let's just revert to the original implementation as there is no improvement for this one.
Sorry for making you work this hard.

Reply:
No problem. I researched a lot, and it seems that there is no viable parallel implementation for 'selection algorithms' as they are highly linear.

I will revert to the original implementation.